design-proposals/klip-33-key-format.md

```diff
@@ -0,0 +1,299 @@
+such solutions tend to increase the complexity of the system, and generally hurt performance.
+
+To open ksqlDB up to new problems spaces and to drive adoption, ksqlDB should support other key formats. 
+
+## What is in scope
+
+ * Addition of a new optional `KEY_FORMAT` property in the `WITH` clause, to set the key format.
+ * Addition of a new optional `FORMAT` property in the `WITH` clause, to set both the key & value formats.
+ * Addition of new server configuration to provide defaults for key and value formats.
+ * Removal of requirement for `VALUE_FORMAT` property in the `WITH` clause of CT/CS statements, where
+   the server configuration provides a default.
+ * Support for additional key column data types, where the key format supports it:
+    * `DECIMAL`
+    * `BOOLEAN`
+ * Support of the following key formats:
+    * `DELIMITED`: single key columns as a single string value.
+    * `JSON` / `JSON_SR`: single key column as an anonymous value, i.e. not within a JSON object.
+    * `AVRO`: single key column as an anonymous value, i.e. not within an Avro record.
+  * Full support of these key formats for all supported SQL syntax.
+  * Automatic repartitioning of streams and tables for joins where key formats do not match.
+  * Support for reading & writing key schemas to & from the schema registry. 
+
+## What is not in scope
+
+ * Support for multiple key columns: this will come later.
+ * Support for single key columns _wrapped_ in an envelope of some kind: this will come later.
+ * Support for complex key column data types, i.e. array, struct and map 
+ * Support for `PROTOBUF` keys, as this requires support for wrapped keys: this will come later.
+
+## Value/Return
+
+We know from customers and community members that there are a lot of people that have data with 
+non-`KAFKA` formatted keys. This is the first step to unlocking that data and use-cases.
+
+With support for `AVRO` and `JSON` key formats there are a lot of existing use-cases that suddenly
+no longer require pre-processing, or tricky Connect SMTs configured and there are new use-cases,
+which ksqlDB was previously unsuitable for, as documented in the motivation section, which it can 
+now handle.
+
+## Public APIS
+
+### CREATE properties
+
+The following new properties will be accepted in the `WITH` clause of `CREATE` statements for streams
+and tables.
+
+* `KEY_FORMAT`: sets the key format, works long the same lines as the existing `VALUE_FORMAT`.
+* `FORMAT`: sets both the key and value format with a single property.
+
+The existing `VALUE_FORMAT` property will no longer be required, if the server config provides a 
+default. (See below).
+
+### Server configs
+
+The following new configuration options will be added. These configurations can be set globally, 
+within the application property file, or locally, via the `SET` command.
+
+* `ksql.persistence.default.format.key`: the default key format.
```

Comment:
Should there also be `ksql.persistence.default.format` ?

Reply:
I don't think so.  If you're setting config, easy enough to just set both.